### PR TITLE
Rework DUID setting

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Comment by keszybz** [comment]
Note that github displays the commits in random order. Proper (reverse) order:
- e0e5f7f9f5 man: add a description of DUIDType and DUIDRawData
- d04670aaaf man: s/similar/similarly/
- eeee8bd5e7 test-networkd-conf: add tests for the parsing functions
- 3d88edb0c8 networkd: rework duid_{type,duid_type,duid,duid_len} setting
- 927eb71368 dh-dhcp{,6}-client: change the semantics of DUID setting
- ffee763f74 dhcp-identifier: un-inline dhcp_validate_duid_len
- fe38ae313a sd-dhcp: constify uint8_t duid[] parameter
- 634f0f983c networkd: rework headers to avoid circular includes

**Comment by keszybz** [comment]
@vinaykul If you could have a look, I'd be grateful.

**Comment by vinaykul** [comment]
@keszybz Thanks for taking care of this and overall this looks clean and good.
We do lose the ability to specify the entire DUID in DUIDRawData with the 'raw' type with the new semantics but it allows for specifying various types in a more user friendly way without ambiguity. So this is headed in a good direction. I'll setup my migration test environment tomorrow and verify the latest changes.

**Comment by keszybz** [comment]
If we need to allow unknown types, we could extend the parse for DUIDType to allow numerical values. But I think that it'd be more interesting to add support for llt and ll first.

**Comment by poettering** [comment]
regarding the circular include thing, see my comments on that other PR, regarding typedef redeclaration...

File: src/libsystemd-network/sd-dhcp-client.c [hunk 1]
```diff
@@ -298,13 +298,32 @@ int sd_dhcp_client_set_client_id(sd_dhcp_client *client, uint8_t type,
...
+        assert_return(duid_len == 0 || duid != NULL, -EINVAL);
 
+        if (duid != NULL) {
+                r = dhcp_validate_duid_len(duid_type, duid_len);
+                if (r < 0)
+                        return r;
+        }
+
+        zero(client->client_id);
         client->client_id.type = 255;
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-client.c:347]: unrelated to the patch at hand, but shouldn't there be a define of some kind fo this value 255?

File: src/libsystemd-network/sd-dhcp-client.c [hunk 2]
```diff
@@ -298,13 +298,32 @@ int sd_dhcp_client_set_client_id(sd_dhcp_client *client, uint8_t type,
...
+/**
+ * Sets IAID and DUID. If duid is non-null, the DUID is set to duid_type + duid
+ * without further modification. Otherwise, if duid_type is supported, DUID
+ * is set based on that type. Otherwise, an error is returned.
+ */
+int sd_dhcp_client_set_iaid_duid(
+                sd_dhcp_client *client,
+                uint32_t iaid,
+                uint16_t duid_type,
+                const uint8_t *duid,
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-client.c]: Hmm, shouldn't this by a "const void*" ptr here? I mean, that's usually what we use when we just want to pass "some data", and it has the benefit that people can pass in a struct of some kind, if they feel like it, without having to cast...

File: src/network/networkd-network.h [hunk 3]
```diff
@@ -68,6 +68,14 @@ typedef enum LLDPMode {
         _LLDP_MODE_INVALID = -1,
 } LLDPMode;
 
+typedef struct DUID {
+        /* Value of Type in [DHCP] section */
+        DUIDType type;
+
+        uint8_t duid_len;
+        uint8_t duid[MAX_DUID_LEN];
+} DUID;
```
Comment (poettering) [inline src/network/networkd-network.h:77]: I think structs called "foo" should not have a field called "foo"... Also, let's not abbreviate unnecessarily in types. Maybe turn this into:

``` c
typedef struct {
        DUIDType type;
        sizet length;
        uint8_t data[DUID_LENGTH_MAX];
} DUID;
```

**Comment by poettering** [comment]
thanks for looking into this. looks much better already!

**Comment by vinaykul** [comment]
@keszybz I ran through my migration scenarios and for IPv4, IPv6 and per-network override and it looks good. However, I realized that the new semantics required me to configure DUID as follows:
screen shot 2016-05-01 at 6 25 40 pm

This is quite different from the interface that @teg initially suggested in https://github.com/systemd/systemd/pull/2761#issuecomment-190835118 where link-layer-time means the user specifies MACAddress and TimeStamp. Looking forward, if we had a facility to specify MACAddress, TimeStamp, and user specifies all these including RawData, we will hit ambiguity on what to make of RawData. I suppose that interface is overly complex to implement due to the ambiguity that it brings in - especially given the constraint that DUIDType can appear in any order. 

Should we just simplify this further and get rid of DUIDType entirely? So we just use DUIDRawData field that holds the DUIDType as well. 

i.e In this case it just becomes: 
DUIDRawData=00:01:00:01:1d:16:c8:f4:02:00:f9:3a:77:7a

instead of:
DUIDType=link-layer-time
DUIDRawData=00:01:1d:16:c8:f4:02:00:f9:3a:77:7a

File: src/libsystemd-network/sd-dhcp-client.c (hunk 1 above)
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp-client.c:347]: There should.

File: src/network/networkd-network.h (hunk 3 above)
Comment (keszybz) [inline src/network/networkd-network.h:77]: I thought about the name too. It was either `duid` or  `data` or `raw_data` to match the config option name. I'll rename it.

File: src/libsystemd-network/sd-dhcp-client.c (hunk 2 above)
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp-client.c]: Hm, possibly. But `const uint8_t *` is standard in sd-dhcp-client: set_mac, set_client_id. We should either convert all such arguments.

**Comment by keszybz** [comment]
> Should we just simplify this further and get rid of DUIDType entirely? So we just use DUIDRawData field that holds the DUIDType as well.

I don't think so. I see DUIDRawData as an escape hatch for unsupported DUID types or experimentation.  Normally users should specify one of the other duid creation mechanisms. The data that is needed for that mechanism will be either provided by the user (e.g. DUIDMACAddress in case of LL or LLT), and a suitable default will be chosen otherwise. I think we should allow adding this metadata in a nice human-readable syntax, and only provide support for specifying byte values directly as a fallback.

It is true that by DUIDRawData and e.g. DUIDMacAddress provide the same data. I think that we can keep the rule that DUIDRawData overrides the other settings. The rule then is that fields appropriate for the specified DUIDType are used, and the rest are ignored, and DUIDRawData trumps other settings.

Namely:

```
DUIDType=verndor
DUIDRawData=1:2:3:...
```

```
DUIDType=link-layer-time
DUIDRawData=1:2:3:...
```

or

```
DUIDType=link-layer-time
DUIDMacAddress=wlp3s0 (or maybe a mac address)
DUIDTimestamp=...
```

etc.

File: src/libsystemd-network/sd-dhcp-client.c (hunk 2 above)
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp-client.c]: Hm, OTOH, we cannot change the type without breaking ABI. So let's "fix" set_duid while we can, and leave the other ones for later.

Comment (poettering) [inline src/libsystemd-network/sd-dhcp-client.c]: but you are right, i think sd-dhcp should really use "const void*" for these too.

**Comment by poettering** [comment]
I think a mechanism where the first assignment that implies a specific DUID type assigns it would make sense too, and then from that point on refuse settings that don't match the right DUID type loudly, with a warning.

i.e. if people just use DUIDRaw=... as first statement, we should derive the DUID type from that. And if people then in the same file use DUIDType= too, and it doesn't match that we complain and say that the configuration makes no sense. Also, if they use any of the other DUIDxyz= settings and the type doens't match that, then we'd complain too.

What is important is that we always get into a well-defined state, and for that two algorithms are acceptable I think:

a) the last setting wins, and possibly overrides and trashes what came before, silently

b) the first setting wins, and all later conflicting settings result in a loud warning

**Comment by keszybz** [comment]
About the typedef redeclaration: I think it is fine to rely on this gcc extension. Of course only in the internal code, not in exported headers. Our code does not compile with a compiler that does not support various gcc extensions, so I don't think it makes sense to worry about this particular aspect, when we use %m, and _cleanup_ and a bunch of other things. I now tested with gcc and clang and they both silently accept the duplicated typedefs. In fact any compiler which supports C++ is likely to support this, since it's legal in C++ and forbidding it doesn't make much sense.

So unless you insist, I'd prefer to keep the patch as is.

**Comment by keszybz** [comment]
After re-reading the description of DUIDs in RFC3315 again, I think that omitting DUIDType and guessing the right value would be too much magic, because there isn't a unique mapping from parameters to DUIDType. E.g. for LL and LLT I don't think the user should specify _anything_, instead we should pick one of the devices and some timestamp, touch a file on disk to make the timestamp permanent, and generate the whole DUID without any further input. If the user wants to override the MAC, that's fine, but they shouldn't have to override also the time. So I think things will be easier to grok if DUIDType is explicit. Also in the RFC, duid type is a separate field.

As an analogy, we don't change service Type to forking just because you specify a PIDFile. That'd be too much magic too.

**Comment by keszybz** [comment]
This addresses all the comments, except the two to which I responded above. Please re-re-review.

**Comment by poettering** [comment]
Hmm, I am generally fine with the typedef thing, but we probably should check starting with which gcc version this ceased to be a warning/error.

**Comment by poettering** [comment]
supposedly C allows redeclarations of typedefs since C11. Now I wonder when gcc started to support that part...

**Comment by keszybz** [comment]
gcc 4.6 [https://bugzilla.redhat.com/show_bug.cgi?id=767538#c2]. 4.6.0 was released in March 2011.

**Comment by teg** [comment]
This looks good now. Thanks a bunch for working on this!

**Pull request merged**